_posts/2019/2019-02-10-be-careful-assigning-to-has-one-relations.md

```diff
@@ -0,0 +1,110 @@
+---
+title: Be Careful Assigning to has_one Relations
+description: Strange Think
+layout: article
+category: ruby
+image:
+  base: '2019/be-careful-assigning-to-has-one-relations'
+  alt: 'Broken net'
+  source: "https://unsplash.com/photos/IV--3UEiHlI"
+  credit: "Ben Hershey"
+---
+
+For much of the time, when building relationships between models you typically use `has_many` and `belongs_to`. There are some circumstances where a `has_one` relationship is more appropriate.
+
+However, the behaviour of `has_one` has some wrinkles in its behaviour that make it a little trickier to deal with. See the side-effects of the generated `#association=` method [in the Rails documentation](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_one).
+
+
+## When…
+
+…assigning an invalid new object means you lose the relationship with the currently related model.
+
+```ruby
+class Biography < ApplicationRecord
+  belongs_to :person, optional: true
+
+  validates :text, presence: true
+end
+
+class Person < ApplicationRecord
+  has_one :biography
+end
+
+person = Person.create(name: "Andy Croll")
+=> #<Person id: 1, name: "Andy Croll" ...>
+person.create_biography(text: "Writes emails")
+=> #<Biography id: 1, person_id: 1, text: "Writes emails" ...>
+person.biography = Biography.new
+=> #<Biography id: nil, person_id: 1, text: nil ...>
+person.reload.biography
+=> nil
+```
+
+Oh my.
+
+
+## Use…
+
+…a transaction to wrap the deletion of the existing object. Only delete the existing related object if the new object is valid.
```

Comment:
You can say this more clearly: Use a transaction to make sure the existing object is deleted only if the new object is valid — or something like that...